train.py

```diff
@@ -312,47 +315,46 @@ def log_step(step_type, device, step, log_output=None, tracker=None):
-        }
-        lr = trainers[devices[0]].get_lr()
-
-        # Load the latest checkpoint if one is available and restore the
-        # corresponding train iterator
-        extra_state, epoch_itr = checkpoint_utils.load_checkpoint(
-            args, trainers[devices[0]])
-        if extra_state is not None:
-            # checkpoint detected, load saved model weights to all devices
-            xu.eprint(
-                'checkpoint detected, device 0 meters need to be '
-                're-loaded to device'
-            )
-            checkpoint_utils.load_checkpoint_tpu(args, trainers, devices[0])
-        valid_subsets = args.valid_subset.split(',')
-        return task, trainers, model_parallel, epoch_itr, lr, valid_subsets
-
-    def train_loop_fn(model, loader, device, context):
-        trainer = trainers[str(device)]
-        stats, log_output = None, None
-        tracker = xm.RateTracker()
+    def prepare_task(args, xla_device):
+       # Setup task, e.g., translation, language modeling, etc.
+       task = tasks.setup_task(args)
+
+       # Load valid dataset (we load training data below, based on the latest checkpoint)
+       for valid_sub_split in args.valid_subset.split(','):
+           task.load_dataset(valid_sub_split, combine=True, epoch=0)
+
+       # Build models and criteria to print some metadata
+       torch.manual_seed(args.seed)
+       model, criterion = task.build_model(args), task.build_criterion(args)
+       xm.master_print(model)
+       xm.master_print('| model {}, criterion {}'.format(
+           args.arch, criterion.__class__.__name__))
+       xm.master_print('| num. model params: {} (num. trained: {})'.format(
+           sum(p.numel() for p in model.parameters()),
+           sum(p.numel() for p in model.parameters() if p.requires_grad)))
+       model = model.to(xla_device)
+       trainer = Trainer(args, task, model, criterion, xla=True)
+       lr = trainer.get_lr()
+
+       # Load the latest checkpoint if one is available and restore the
+       # corresponding train iterator
+       # set distributed args here to shard data
+       trainer.args.distributed_rank = xm.get_ordinal()
+       trainer.args.distributed_world_size = xm.xrt_world_size()
+       extra_state, epoch_itr = checkpoint_utils.load_checkpoint(args, trainer)
+       trainer.args.distributed_rank = 0
+       trainer.args.distributed_world_size = 1
+       valid_subsets = args.valid_subset.split(',')
+       ordinal = xm.get_ordinal(defval=-1)
+       device_str = str(xla_device) if ordinal < 0 else '{}/{}'.format(xla_device, ordinal)
+       return task, trainer, model, epoch_itr, lr, valid_subsets, device_str
+
+    def train_loop_fn(device, trainer, loader):
+        stats, log_output, tracker = None, None, xm.RateTracker()
         for i, samples in enumerate(loader):
+            if i+1 == len(loader._loader._loader):
+                #FIXME: last batches are incomplete
```

Comment:
Is this meant to be here?

Reply:
I asked Davide above, my intention is to get rid of this, but we need to drop the last batch safely somewhere. This code, as it stands, is actually buggy, it would hang if one of the loaders produces 100 batches and the others produce 101.

Reply:
The parallel loader should be already handling this.
Was fixed long time ago.
Now before releasing a batch, wait that we have N-devices worth of them.

Reply:
> Now before releasing a batch, wait that we have N-devices worth of them.

Not sure I understand the last bit. Are you stating that's what happens or are you requesting a change?

Reply:
What I am saying os that the parallel loader will have release an N-device batch once it is full:

https://github.com/pytorch/xla/blob/6bacd3ad2b3a8fdcd94cbdae96e40d54b409e22a/torch_xla/distributed/parallel_loader.py#L154

So if you have special handling for that, you should be able to drop it.

Reply:
So I added this:

```python
         for i, samples in enumerate(loader):
+            print(device, i, len(loader._loader._loader), samples[0]['net_input']['src_tokens'].shape, f
lush=True)
+            continue
             if i+1 == len(loader._loader._loader):
                 #FIXME: last batches are incomplete
                 break
```

and then I run and get different length iterators on different devices:

```bash
(torch-xla-nightly)  % ./QUICK-fst.sh 2> /dev/null | grep torch > tmp
(torch-xla-nightly)  % grep xla tmp | grep ' 2[0-9] 25' | grep 24
xla:0/1 24 25 torch.Size([256, 64])
xla:0/2 24 25 torch.Size([256, 64])
xla:1/0 24 25 torch.Size([256, 64])
(torch-xla-nightly)  % grep xla tmp | grep ' 2[0-9] 25' | grep 23
xla:0/1 23 25 torch.Size([256, 64])
xla:0/6 23 25 torch.Size([256, 64])
xla:0/2 23 25 torch.Size([256, 64])
xla:0/4 23 25 torch.Size([256, 64])
xla:0/5 23 25 torch.Size([256, 64])
xla:0/3 23 25 torch.Size([256, 64])
xla:1/0 23 25 torch.Size([256, 64])
xla:0/7 23 25 torch.Size([256, 64])
```

Reply:
Ohhh .... right!
With MP there is a `local view`, and the PyTorch `torch.utils.data.distributed.DistributedSampler` does the things.
So it could actually feed us inconsistent number of batches for each replica.

https://github.com/pytorch/pytorch/blob/06bb74ce96f80856ded884282c17777f61fd3212/torch/utils/data/distributed.py#L39

I think `ceil()` is wring there ...

Reply:
Oookay so I know what is going on here.

* In loop 25, Fairseq's iterators produce full batches for processes 0-2 and empty batches for processes 3-7.
* loop 25 consequently errors with 'KeyError: 'net_input'`

So it looks like I need to manually remove the last batch just like I was doing. I will do that a bit more properly and update this pr soon.

Reply:
To me looks like the sampler is wrong. IIRC we talked about that very same line of code before ...

Reply:
If I understand correctly, since they don't care about devices doing different things etc, they use ceil in order to not drop any batches. It doesn't work well with our use case, so I need to handle the last batch.

Reply:
IMHO it makes sense to post a patch for it.
We don't want to patch every dataloader we use in every model we embark.

Reply:
So what do we do?
We go with your custom handling, post a PR, then come back to fix this?
Are you taking care of the PR to pytorch?

Reply:
Given that the pytorch pr may take long, I'd like to check this in and come back to it. I'll take care of the pytorch pr.